File: Pythoncode/32.最长有效括号.py
class Solution:
    """
    https://leetcode-cn.com/problems/longest-valid-parentheses/solution/zui-chang-you-xiao-gua-hao-by-powcai/
    """
    def longestValidParentheses(self, s: str) -> int:
        if  len(s) == 0:
            return 0
        res = []
        stack = []
        for i in range(len(s)):
            if s[i] == "(":
                stack.append(i)
            else:
                if len(stack) != 0 and s[stack[-1]] == "(":
                    res.append(stack.pop())
                    res.append(i)
                else:
                    stack.append(i)
        res.sort()
        i = 0
        maxlen = 0
        while i < len(res):
            j = i
            while j < len(res) and res[j] - res[i] == j - i:
                maxlen = max(maxlen, j - i + 1)
                j += 1
            i = j
        return maxlen

File: Cppcode/32.最长有效括号.cpp
#include <string>
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

class Solution {
public:
    int longestValidParentheses(string s) {
        if (s == "") return 0;
        stack<decltype(s.size())> stk; // 里面放的都是 s 的index
        vector<int> v = vector<int> (s.size(), 0); // v[j] 记录的是以 s[j] 作为结尾的最长括号序列的长度
        for (decltype(s.size()) j = 0; j < s.size(); j++)
        {
            if (s[j] == '(')
            {
                stk.push(j);
            }
            else if (stk.size() != 0 and s[stk.top()] == '(')
            {
                auto i = stk.top();
                stk.pop();
                if (i == 0) v[j] = j - i + 1;
                else v[j] = j - i + 1 + v[i-1];
            }
            else
            {
                ;
            }
        }
        return *max_element(v.begin(), v.end());    
    }
};